service/storage_proxy.cc

```diff
@@ -127,20 +127,26 @@ seastar::metrics::label_instance current_scheduling_group_label() {
 
 }
 
-template<typename ResultTuple, typename SourceTuple>
-static future<ResultTuple> encode_replica_exception_for_rpc(gms::feature_service& features, future<SourceTuple>&& f) {
-    if (!f.failed()) {
-        return make_ready_future<ResultTuple>(utils::tuple_insert<ResultTuple>(f.get(), replica::exception_variant{}));
-    }
-    std::exception_ptr eptr = f.get_exception();
+template<typename ResultTuple>
+static future<ResultTuple> encode_replica_exception_for_rpc(gms::feature_service& features, std::exception_ptr eptr) {
     if (features.typed_errors_in_read_rpc) {
         if (auto ex = replica::try_encode_replica_exception(eptr); ex) {
-            return make_ready_future<ResultTuple>(utils::tuple_insert<ResultTuple>(utils::make_default_rpc_tuple<SourceTuple>(), std::move(ex)));
+            ResultTuple encoded_ex = utils::make_default_rpc_tuple<ResultTuple>();
```

Comment:
I would add a `static_assert` or a constraint here that `ResultTuple` contains `replica::exception_variant`, so we can be sure in compilation time that `std::get` below doesn't throw.

Reply:
Note that the previously used `tuple_insert` did such checks (`constexpr auto match =...` followed by `static_assert(match.second == 1, ...)`)

Reply:
https://en.cppreference.com/w/cpp/utility/tuple/get
> 5-8) Extracts the element of the tuple `t` whose type is T. Fails to compile unless the tuple has exactly one element of that type.

We could still consider `static_assert` here so that we get a potentially nicer error, but I'm not sure if it's worth it.